// @zerodevapp/sdk/dist/src/paymasters/VerifyingPaymasterAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifyingPaymasterAPI = void 0;
const properties_1 = require("@ethersproject/properties");
const api_1 = require("../api");
const errors_1 = require("../errors");
const PaymasterAPI_1 = require("./PaymasterAPI");
const utils_1 = require("../utils");
class VerifyingPaymasterAPI extends PaymasterAPI_1.PaymasterAPI {
    constructor(projectId, paymasterUrl, chainId, entryPointAddress, paymasterProvider) {
        super();
        this.projectId = projectId;
        this.paymasterUrl = paymasterUrl;
        this.chainId = chainId;
        this.entryPointAddress = entryPointAddress;
        this.paymasterProvider = paymasterProvider;
    }
    async getPaymasterResp(userOp) {
        const resolvedUserOp = await (0, properties_1.resolveProperties)(userOp);
        const hexifiedUserOp = (0, utils_1.hexifyUserOp)(resolvedUserOp);
        const paymasterResp = await (0, api_1.signUserOp)(this.projectId, this.chainId, hexifiedUserOp, this.entryPointAddress, this.paymasterUrl, this.paymasterProvider);
        if (paymasterResp === undefined) {
            throw errors_1.ErrTransactionFailedGasChecks;
        }
        return paymasterResp;
    }
}
exports.VerifyingPaymasterAPI = VerifyingPaymasterAPI;
