// @zerodevapp/contracts/types/factories/IPolicy__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type { IPolicy, IPolicyInterface } from "../IPolicy";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "to",
        type: "address",
      },
      {
        internalType: "uint256",
        name: "value",
        type: "uint256",
      },
      {
        internalType: "bytes",
        name: "data",
        type: "bytes",
      },
      {
        internalType: "enum Enum.Operation",
        name: "operation",
        type: "uint8",
      },
    ],
    name: "executeAndRevert",
    outputs: [
      {
        internalType: "bool",
        name: "",
        type: "bool",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

export class IPolicy__factory {
  static readonly abi = _abi;
  static createInterface(): IPolicyInterface {
    return new utils.Interface(_abi) as IPolicyInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IPolicy {
    return new Contract(address, _abi, signerOrProvider) as IPolicy;
  }
}
