// @zerodevapp/kernel-contracts-v2/types/KillSwitchAction.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type { FunctionFragment, Result } from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "./common";

export interface KillSwitchActionInterface extends utils.Interface {
  functions: {
    "killSwitchValidator()": FunctionFragment;
    "toggleKillSwitch()": FunctionFragment;
  };

  getFunction(
    nameOrSignatureOrTopic: "killSwitchValidator" | "toggleKillSwitch"
  ): FunctionFragment;

  encodeFunctionData(
    functionFragment: "killSwitchValidator",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "toggleKillSwitch",
    values?: undefined
  ): string;

  decodeFunctionResult(
    functionFragment: "killSwitchValidator",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "toggleKillSwitch",
    data: BytesLike
  ): Result;

  events: {};
}

export interface KillSwitchAction extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: KillSwitchActionInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    killSwitchValidator(overrides?: CallOverrides): Promise<[string]>;

    toggleKillSwitch(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  killSwitchValidator(overrides?: CallOverrides): Promise<string>;

  toggleKillSwitch(
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    killSwitchValidator(overrides?: CallOverrides): Promise<string>;

    toggleKillSwitch(overrides?: CallOverrides): Promise<void>;
  };

  filters: {};

  estimateGas: {
    killSwitchValidator(overrides?: CallOverrides): Promise<BigNumber>;

    toggleKillSwitch(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    killSwitchValidator(
      overrides?: CallOverrides
    ): Promise<PopulatedTransaction>;

    toggleKillSwitch(
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
