// @zerodevapp/kernel-contracts-v2/dist/types/factories/IAddressBook__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IAddressBook__factory = void 0;
var ethers_1 = require("ethers");
var _abi = [
    {
        inputs: [],
        name: "getOwners",
        outputs: [
            {
                internalType: "address[]",
                name: "",
                type: "address[]",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
var IAddressBook__factory = exports.IAddressBook__factory = /** @class */ (function () {
    function IAddressBook__factory() {
    }
    IAddressBook__factory.createInterface = function () {
        return new ethers_1.utils.Interface(_abi);
    };
    IAddressBook__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    IAddressBook__factory.abi = _abi;
    return IAddressBook__factory;
}());
