// @zerodevapp/sdk/dist/src/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC20_APPROVAL_AMOUNT = exports.PEPE_ADDRESS = exports.USDC_ADDRESS = exports.ERC1155_ABI = exports.ERC20_ABI = exports.ERC721_ABI = exports.CHAIN_ID_TO_NODE = exports.DEFAULT_MAX_TX_RETRIES = exports.WALLET_CONNECT_RELAY_URL = exports.WALLET_CONNECT_PROJECT_ID = exports.INFURA_API_KEY = exports.UPDATE_SINGLETON_ADDRESS = exports.ENTRYPOINT_ADDRESS = exports.BUNDLER_URL = exports.PAYMASTER_ABI = exports.PAYMASTER_URL = exports.LOGGER_URL = exports.BACKEND_URL = void 0;
const ethers_1 = require("ethers");
exports.BACKEND_URL = (_a = process.env.REACT_APP_ZERODEV_BACKEND_URL) !== null && _a !== void 0 ? _a : 'https://backend-vikp.onrender.com';
exports.LOGGER_URL = (_b = process.env.REACT_APP_ZERODEV_LOGGER_URL) !== null && _b !== void 0 ? _b : 'https://prod-logger.onrender.com';
exports.PAYMASTER_URL = (_c = process.env.REACT_APP_ZERODEV_PAYMASTER_URL) !== null && _c !== void 0 ? _c : 'https://v0-6-paymaster.onrender.com';
var paymaster_abi_json_1 = require("./abi/paymaster_abi.json");
Object.defineProperty(exports, "PAYMASTER_ABI", { enumerable: true, get: function () { return __importDefault(paymaster_abi_json_1).default; } });
exports.BUNDLER_URL = (_d = process.env.REACT_APP_ZERODEV_BUNDLER_URL) !== null && _d !== void 0 ? _d : 'https://v0-6-meta-bundler.onrender.com';
exports.ENTRYPOINT_ADDRESS = '0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789';
exports.UPDATE_SINGLETON_ADDRESS = '0x3d4d0cab438cee791b7405cf10448daaa98087c0';
exports.INFURA_API_KEY = 'f36f7f706a58477884ce6fe89165666c';
exports.WALLET_CONNECT_PROJECT_ID = '9832ea3eefe6c1b75a689ed0c90ce085';
exports.WALLET_CONNECT_RELAY_URL = 'wss://relay.walletconnect.com';
exports.DEFAULT_MAX_TX_RETRIES = 3;
exports.CHAIN_ID_TO_NODE = {
    1: `https://mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    5: `https://goerli.infura.io/v3/${exports.INFURA_API_KEY}`,
    137: `https://polygon-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    80001: `https://polygon-mumbai.infura.io/v3/${exports.INFURA_API_KEY}`,
    10: `https://optimism-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    420: `https://optimism-goerli.infura.io/v3/${exports.INFURA_API_KEY}`,
    42161: `https://arbitrum-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    // 42161: 'https://evocative-stylish-dinghy.arbitrum-mainnet.discover.quiknode.pro/80b526d14fa9fd9a8b0db1e65554acaf00c6a1ab/',
    421613: `https://arbitrum-goerli.infura.io/v3/${exports.INFURA_API_KEY}`,
    43114: `https://avalanche-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    43113: `https://avalanche-fuji.infura.io/v3/${exports.INFURA_API_KEY}`,
    1313161554: `https://aurora-mainnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    1313161555: `https://aurora-testnet.infura.io/v3/${exports.INFURA_API_KEY}`,
    56: 'https://wandering-quaint-reel.bsc.quiknode.pro/508c3d245c14adb8689ed4073d29aa5795dfa24e/',
    97: 'https://sly-indulgent-paper.bsc-testnet.quiknode.pro/ab7e00c229f5967334160958e40fd6a4d893fb93/',
    84531: 'https://icy-long-mountain.base-goerli.quiknode.pro/5b80d93e97cc9412a63c10a30841869abbef9596/'
};
exports.ERC721_ABI = [
    'function transferFrom(address from, address to, uint256 tokenId) external',
    'function safeTransferFrom(address from, address to, uint256 tokenId) external',
    'function safeTransferFrom(address from, address to, uint256 tokenId, bytes calldata _data) external',
    'function balanceOf(address owner) external view returns (uint256 balance)'
];
exports.ERC20_ABI = [
    'function transfer(address to, uint256 value) external returns (bool)',
    'function transferFrom(address from, address to, uint256 value) external returns (bool)',
    'function approve(address spender, uint256 value) external returns (bool)',
    'function allowance(address owner, address spender) external view returns (uint256)',
    'function balanceOf(address owner) external view returns (uint256)'
];
exports.ERC1155_ABI = [
    'function safeTransferFrom(address from, address to, uint256 id, uint256 amount, bytes calldata data) external',
    'function safeBatchTransferFrom(address from, address to, uint256[] calldata ids, uint256[] calldata amounts, bytes calldata data) external',
    'function balanceOf(address account, uint256 id) external view returns (uint256)',
    'function balanceOfBatch(address[] calldata accounts, uint256[] calldata ids) external view returns (uint256[] memory)'
];
exports.USDC_ADDRESS = {
    1: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
    // 5: '0x2f3A40A3db8a7e3D09B0adfEfbCe4f6F81927557',
    137: '0x2791Bca1f2de4661ED88A30C99A7a9449Aa84174',
    // 80001: '',
    // 420: '0x7F5c764cBc14f9669B88837ca1490cCa17c31607',
    42161: '0xFF970A61A04b1cA14834A43f5dE4533eBDDB5CC8',
    // 421613: '',
    43114: '0xB97EF9Ef8734C71904D8002F8b6Bc66Dd9c48a6E'
    // 43113: '',
    // 56: '',
    // 97: ''
};
exports.PEPE_ADDRESS = {
    1: '0x6982508145454Ce325dDbE47a25d4ec3d2311933'
};
exports.ERC20_APPROVAL_AMOUNT = {
    // ETH
    [exports.USDC_ADDRESS[1]]: ethers_1.ethers.utils.parseUnits('100', 6),
    // Goerli
    [exports.USDC_ADDRESS[5]]: ethers_1.ethers.utils.parseUnits('100', 6),
    // Polygon
    [exports.USDC_ADDRESS[137]]: ethers_1.ethers.utils.parseUnits('10', 6),
    // Arbitrum
    [exports.USDC_ADDRESS[42161]]: ethers_1.ethers.utils.parseUnits('10', 6),
    // Avalanche
    [exports.USDC_ADDRESS[43114]]: ethers_1.ethers.utils.parseUnits('10', 6),
    // ETH
    [exports.PEPE_ADDRESS[1]]: ethers_1.ethers.utils.parseUnits('50000000'),
    '0x3870419Ba2BBf0127060bCB37f69A1b1C090992B': ethers_1.ethers.utils.parseUnits('100', 18)
};
