// @zerodevapp/sdk/dist/src/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrNoIdentifierProvided = exports.ErrTransactionFailedGasChecks = exports.ErrTransactionRejectedByUser = exports.ErrUnsupportedIdentity = exports.ErrUnsupportedNetwork = exports.ErrSDKNotInitialized = void 0;
const ErrSDKNotInitialized = Error('SDK not initialized');
exports.ErrSDKNotInitialized = ErrSDKNotInitialized;
const ErrUnsupportedNetwork = Error('Unsupported network');
exports.ErrUnsupportedNetwork = ErrUnsupportedNetwork;
const ErrUnsupportedIdentity = Error('Unsupported identity');
exports.ErrUnsupportedIdentity = ErrUnsupportedIdentity;
const ErrTransactionRejectedByUser = Error('Transaction rejected by user');
exports.ErrTransactionRejectedByUser = ErrTransactionRejectedByUser;
const ErrTransactionFailedGasChecks = Error('Transaction failed gas checks');
exports.ErrTransactionFailedGasChecks = ErrTransactionFailedGasChecks;
const ErrNoIdentifierProvided = Error('No identity token, private key, or Web3 provider was provided');
exports.ErrNoIdentifierProvided = ErrNoIdentifierProvided;
