// @zerodevapp/sdk/dist/src/KernelAccountAPI.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KernelAccountAPI = void 0;
const ethers_1 = require("ethers");
const utils_1 = require("ethers/lib/utils");
const BaseAccountAPI_1 = require("./BaseAccountAPI");
const multisend_1 = require("./multisend");
const contracts_new_1 = require("@zerodevapp/contracts-new");
const utils_2 = require("./utils");
/**
 * An implementation of the BaseAccountAPI using Gnosis Safe.
 * - Pass "owner" address and "index" nonce to the factory
 * - owner signs requests using normal "Ethereum Signed Message" (ether's signer.signMessage())
 * - nonce is a public variable "nonce"
 * - execute method is "execTransactionFromModule()", since the entrypoint is set as a module
 */
class KernelAccountAPI extends BaseAccountAPI_1.BaseAccountAPI {
    constructor(params) {
        super(params);
        this.factoryAddress = params.factoryAddress;
        this.owner = params.owner;
    }
    async _getAccountContract() {
        if (this.accountContract == null) {
            this.accountContract = contracts_new_1.Kernel__factory.connect(await this.getAccountAddress(), this.provider);
        }
        return this.accountContract;
    }
    /**
     * return the value to put into the "initCode" field, if the account is not yet deployed.
     * this value holds the "factory" address, followed by this account's information
     */
    async getAccountInitCode() {
        const factoryAddr = await this.getFactoryAddress();
        const factoryInitCode = await this.getFactoryAccountInitCode();
        return (0, utils_1.hexConcat)([
            factoryAddr,
            factoryInitCode
        ]);
    }
    async getFactoryAddress() {
        if (this.factoryAddress != null) {
            return this.factoryAddress;
        }
        throw new Error('no factory address');
    }
    async getFactoryAccountInitCode() {
        const ownerAddress = await this.owner.getAddress();
        if (this.factory == null) {
            if (this.factoryAddress != null && this.factoryAddress !== '') {
                this.factory = contracts_new_1.KernelFactory__factory.connect(this.factoryAddress, this.provider);
            }
            else {
                throw new Error('no factory to get initCode');
            }
        }
        return this.factory.interface.encodeFunctionData('createAccount', [ownerAddress, this.index]);
    }
    async getNonce() {
        if (await this.checkAccountPhantom()) {
            return ethers_1.BigNumber.from(0);
        }
        const accountContract = await this._getAccountContract();
        return await accountContract['getNonce()']();
    }
    /**
     * encode a method call from entryPoint to our contract
     * @param target
     * @param value
     * @param data
     */
    async encodeExecute(target, value, data) {
        const accountContract = await this._getAccountContract();
        // the executeAndRevert method is defined on the manager
        const managerContract = contracts_new_1.Kernel__factory.connect(accountContract.address, accountContract.provider);
        return managerContract.interface.encodeFunctionData('executeAndRevert', [
            target,
            value,
            data,
            0
        ]);
    }
    /**
     * encode a method call from entryPoint to our contract
     * @param target
     * @param value
     * @param data
     */
    async encodeExecuteDelegate(target, value, data) {
        const accountContract = await this._getAccountContract();
        // the executeAndRevert method is defined on the manager
        const managerContract = contracts_new_1.Kernel__factory.connect(accountContract.address, accountContract.provider);
        return managerContract.interface.encodeFunctionData('executeAndRevert', [
            target,
            value,
            data,
            1
        ]);
    }
    /**
     * encode a method call from entryPoint to our contract
     * @param target
     * @param value
     * @param data
     */
    async decodeExecuteDelegate(data) {
        const accountContract = await this._getAccountContract();
        // the executeAndRevert method is defined on the manager
        const managerContract = contracts_new_1.Kernel__factory.connect(accountContract.address, accountContract.provider);
        return managerContract.interface.decodeFunctionData('executeAndRevert', data);
    }
    async encodeExecuteBatch(calls) {
        const multiSend = new ethers_1.Contract((0, multisend_1.getMultiSendAddress)(), [
            'function multiSend(bytes memory transactions)'
        ]);
        const multiSendCalldata = multiSend.interface.encodeFunctionData('multiSend', [(0, multisend_1.encodeMultiSend)(calls)]);
        return await this.encodeExecuteDelegate(multiSend.address, 0, multiSendCalldata);
    }
    async signUserOpHash(userOpHash) {
        return await this.owner.signMessage((0, utils_1.arrayify)(userOpHash));
    }
    async signMessage(message) {
        const dataHash = ethers_1.ethers.utils.arrayify(ethers_1.ethers.utils.hashMessage(message));
        let sig = (0, utils_2.fixSignedData)(await this.owner.signMessage(dataHash));
        // If the account is undeployed, use ERC-6492
        if (await this.checkAccountPhantom()) {
            const coder = new ethers_1.ethers.utils.AbiCoder();
            sig = coder.encode(['address', 'bytes', 'bytes'], [
                await this.getFactoryAddress(),
                await this.getFactoryAccountInitCode(),
                sig
            ]) + '6492649264926492649264926492649264926492649264926492649264926492'; // magic suffix
        }
        return sig;
    }
}
exports.KernelAccountAPI = KernelAccountAPI;
