// @zerodevapp/contracts/types/UpdateSingleton.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumber,
  BytesLike,
  CallOverrides,
  ContractTransaction,
  Overrides,
  PopulatedTransaction,
  Signer,
  utils,
} from "ethers";
import type {
  FunctionFragment,
  Result,
  EventFragment,
} from "@ethersproject/abi";
import type { Listener, Provider } from "@ethersproject/providers";
import type {
  TypedEventFilter,
  TypedEvent,
  TypedListener,
  OnEvent,
  PromiseOrValue,
} from "./common";

export interface UpdateSingletonInterface extends utils.Interface {
  functions: {
    "self()": FunctionFragment;
    "update(address)": FunctionFragment;
  };

  getFunction(nameOrSignatureOrTopic: "self" | "update"): FunctionFragment;

  encodeFunctionData(functionFragment: "self", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "update",
    values: [PromiseOrValue<string>]
  ): string;

  decodeFunctionResult(functionFragment: "self", data: BytesLike): Result;
  decodeFunctionResult(functionFragment: "update", data: BytesLike): Result;

  events: {
    "ChangedMasterCopy(address)": EventFragment;
  };

  getEvent(nameOrSignatureOrTopic: "ChangedMasterCopy"): EventFragment;
}

export interface ChangedMasterCopyEventObject {
  singleton: string;
}
export type ChangedMasterCopyEvent = TypedEvent<
  [string],
  ChangedMasterCopyEventObject
>;

export type ChangedMasterCopyEventFilter =
  TypedEventFilter<ChangedMasterCopyEvent>;

export interface UpdateSingleton extends BaseContract {
  connect(signerOrProvider: Signer | Provider | string): this;
  attach(addressOrName: string): this;
  deployed(): Promise<this>;

  interface: UpdateSingletonInterface;

  queryFilter<TEvent extends TypedEvent>(
    event: TypedEventFilter<TEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TEvent>>;

  listeners<TEvent extends TypedEvent>(
    eventFilter?: TypedEventFilter<TEvent>
  ): Array<TypedListener<TEvent>>;
  listeners(eventName?: string): Array<Listener>;
  removeAllListeners<TEvent extends TypedEvent>(
    eventFilter: TypedEventFilter<TEvent>
  ): this;
  removeAllListeners(eventName?: string): this;
  off: OnEvent<this>;
  on: OnEvent<this>;
  once: OnEvent<this>;
  removeListener: OnEvent<this>;

  functions: {
    self(overrides?: CallOverrides): Promise<[string]>;

    update(
      targetSingleton: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<ContractTransaction>;
  };

  self(overrides?: CallOverrides): Promise<string>;

  update(
    targetSingleton: PromiseOrValue<string>,
    overrides?: Overrides & { from?: PromiseOrValue<string> }
  ): Promise<ContractTransaction>;

  callStatic: {
    self(overrides?: CallOverrides): Promise<string>;

    update(
      targetSingleton: PromiseOrValue<string>,
      overrides?: CallOverrides
    ): Promise<void>;
  };

  filters: {
    "ChangedMasterCopy(address)"(
      singleton?: null
    ): ChangedMasterCopyEventFilter;
    ChangedMasterCopy(singleton?: null): ChangedMasterCopyEventFilter;
  };

  estimateGas: {
    self(overrides?: CallOverrides): Promise<BigNumber>;

    update(
      targetSingleton: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<BigNumber>;
  };

  populateTransaction: {
    self(overrides?: CallOverrides): Promise<PopulatedTransaction>;

    update(
      targetSingleton: PromiseOrValue<string>,
      overrides?: Overrides & { from?: PromiseOrValue<string> }
    ): Promise<PopulatedTransaction>;
  };
}
