// @zerodevapp/contracts/dist/types/factories/IPolicy__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
exports.__esModule = true;
exports.IPolicy__factory = void 0;
var ethers_1 = require("ethers");
var _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "to",
                type: "address"
            },
            {
                internalType: "uint256",
                name: "value",
                type: "uint256"
            },
            {
                internalType: "bytes",
                name: "data",
                type: "bytes"
            },
            {
                internalType: "enum Enum.Operation",
                name: "operation",
                type: "uint8"
            },
        ],
        name: "executeAndRevert",
        outputs: [
            {
                internalType: "bool",
                name: "",
                type: "bool"
            },
        ],
        stateMutability: "view",
        type: "function"
    },
];
var IPolicy__factory = /** @class */ (function () {
    function IPolicy__factory() {
    }
    IPolicy__factory.createInterface = function () {
        return new ethers_1.utils.Interface(_abi);
    };
    IPolicy__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    IPolicy__factory.abi = _abi;
    return IPolicy__factory;
}());
exports.IPolicy__factory = IPolicy__factory;
