// @zerodevapp/contracts/dist/types/factories/IProxyCreationCallback__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
exports.__esModule = true;
exports.IProxyCreationCallback__factory = void 0;
var ethers_1 = require("ethers");
var _abi = [
    {
        inputs: [
            {
                internalType: "contract GnosisSafeProxy",
                name: "proxy",
                type: "address"
            },
            {
                internalType: "address",
                name: "_singleton",
                type: "address"
            },
            {
                internalType: "bytes",
                name: "initializer",
                type: "bytes"
            },
            {
                internalType: "uint256",
                name: "saltNonce",
                type: "uint256"
            },
        ],
        name: "proxyCreated",
        outputs: [],
        stateMutability: "nonpayable",
        type: "function"
    },
];
var IProxyCreationCallback__factory = /** @class */ (function () {
    function IProxyCreationCallback__factory() {
    }
    IProxyCreationCallback__factory.createInterface = function () {
        return new ethers_1.utils.Interface(_abi);
    };
    IProxyCreationCallback__factory.connect = function (address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    };
    IProxyCreationCallback__factory.abi = _abi;
    return IProxyCreationCallback__factory;
}());
exports.IProxyCreationCallback__factory = IProxyCreationCallback__factory;
